kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -307,6 +307,38 @@ private static List<Tuple2<List<String>, StructField>> recurseIntoComplexTypes(
     return filtered;
   }
 
+  /* Compute the SchemaChanges using field IDs */
+  static SchemaChanges computeSchemaChangesById(
+      Map<Long, StructField> currentFieldIdToField, Map<Long, StructField> updatedFieldIdToField) {
+    SchemaChanges.Builder schemaDiff = SchemaChanges.builder();
+    for (Map.Entry<Long, StructField> fieldInUpdatedSchema : updatedFieldIdToField.entrySet()) {
+      StructField existingField = currentFieldIdToField.get(fieldInUpdatedSchema.getKey());
+      StructField updatedField = fieldInUpdatedSchema.getValue();
+      // New field added
+      if (existingField == null) {
+        schemaDiff.withAddedField(updatedField);
+      } else if (isUpdatedField(existingField, fieldInUpdatedSchema.getValue())) {
+        // Field renamed or changed types
+        schemaDiff.withUpdatedField(existingField, fieldInUpdatedSchema.getValue());
+      }
+    }
+
+    for (Map.Entry<Long, StructField> entry : currentFieldIdToField.entrySet()) {
+      StructField fieldInUpdatedSchema = updatedFieldIdToField.get(entry.getKey());
+      if (fieldInUpdatedSchema == null) {
+        schemaDiff.withRemovedField(entry.getValue());
+      }
+    }
+
+    return schemaDiff.build();
+  }
+
+  // Returns if the field has been renamed, if the types are different or if fields have been moved
+  private static boolean isUpdatedField(StructField existingField, StructField updatedField) {
+    return !existingField.getName().equals(updatedField.getName())
+        || !existingField.equals(updatedField);
```

Comment:
Doesn't `existingField.equals(updatedField)` check `existingField.getName().equals(updatedField.getName())` by nature?

I think maybe let's think a bit about what we want to consider an update. The 4 fields for a struct field are nullability, (logical) name, dataType,  and metadata. I think we definitely want to consider any change in nullability, (logical) name, dataType to be an update. I think we probably also want to include any change in metadata (since this could include something like a change in generation expression).

Reply:
So I think maybe `existingField.equals(updatedField)` alone is sufficient? Can you clarify the docs in `SchemaChanges` to state an updated field can be due to a change in any of those 4 items?